src/bridge_topic.c

```diff
@@ -126,18 +162,21 @@ int bridge__add_topic(struct mosquitto__bridge *bridge, const char *topic, enum
 		return MOSQ_ERR_INVAL;
 	}
 
+	if(bridge__find_topic(bridge, topic, direction, qos, local_prefix, remote_prefix) != NULL){
+		log__printf(NULL, MOSQ_LOG_INFO, "Duplicate bridge topic '%s', skipping", topic);
+		return MOSQ_ERR_SUCCESS;
+	}
 
 	bridge->topic_count++;
-	topics = mosquitto__realloc(bridge->topics,
-				sizeof(struct mosquitto__bridge_topic)*(size_t)bridge->topic_count);
-
-	if(topics == NULL){
+	cur_topic = mosquitto__malloc(sizeof(struct mosquitto__bridge_topic));
+	if(cur_topic == NULL){
 		log__printf(NULL, MOSQ_LOG_ERR, "Error: Out of memory.");
 		return MOSQ_ERR_NOMEM;
 	}
-	bridge->topics = topics;
+	cur_topic->next = NULL;
+
+	LL_APPEND(bridge->topics, cur_topic);
```

Comment:
I think it would be better to append cur_topic to the list only once it is completely ready at the end of the function. With this as it is, we malloc cur_topic, then add it to the list, then if we get a strdup failure later on then we return with uninitialised memory in some of the struct members. Putting it at the end means if there is a failure we make sure we clean everything up and nothing bad goes on to the list.

Reply:
Strong agree on only adding a fully initialised struct. Looking at it again, there are opportunities to leak memory which is assigned to cur_topic->topic and/or cur_topic->local_prefix if either of the bridge__create_prefix calls fail. Admittedly this is in the case of OOM, so not super dangerous, but incorrect nonetheless. Also more important when bridge__add_topic can be called during the brokers life and not just at config time.

Reply:
I have modified bridge__add_topic to handle OOM cleanup in labelled error blocks at the end of the function. Heavy use of goto, which may not be to your taste. Let me know what you think.